File: simplified-books-core-ng/src/main/java/org/nypl/simplified/books/profiles/ProfileID.java
package org.nypl.simplified.books.profiles;

import com.google.auto.value.AutoValue;

/**
 * A unique profile identifier.
 */

@AutoValue
public abstract class ProfileID implements Comparable<ProfileID> {

  ProfileID() {

  }

  /**
   * Construct a profile identifier.
   *
   * @param id A non-negative integer
   */

  public static ProfileID create(int id) {
    if (id < 0) {
      throw new IllegalArgumentException("Profile identifiers must be non-negative");
    }
    return new AutoValue_ProfileID(id);
  }

  /**
   * @return The raw identifier value
   */

  public abstract int id();

  @Override
  public int compareTo(ProfileID other) {
    return Integer.compare(this.id(), other.id());
  }
}

File: simplified-profiles-api/src/main/java/org/nypl/simplified/profiles/api/ProfileID.kt
package org.nypl.simplified.profiles.api

import java.io.Serializable
import java.util.UUID

/**
 * A unique identifier for a profile.
 */

data class ProfileID(
  val uuid: UUID
) : Serializable, Comparable<ProfileID> {

  override fun compareTo(other: ProfileID): Int =
    this.uuid.compareTo(other.uuid)

  override fun toString(): String =
    this.uuid.toString()

  companion object {

    /**
     * Generate a random profile ID.
     */

    fun generate(): ProfileID =
      ProfileID(UUID.randomUUID())
  }
}
